### PR TITLE
Clehner/maximize views

### LINKED ISSUE #51: Maximization of views
# Feature request

views inside a dashboard should be maximizable, such that a view temporarily takes up all the available space inside the root layout container such that other views are masked. the maximized view should be "minimized" (another word for the title would be good, since "minimize" sounds like it should take up minimal space as in an OS) again to take up its original place and space.

possible icons:
expand: https://fontawesome.com/v4.7.0/icon/expand/
contract / compress: https://fontawesome.com/v4.7.0/icon/compress/

### PR DESCRIPTION
**Pull request by lehnerchristian**
closes #51 

maximize_views

### CODE REVIEW

**Comment by lehnerchristian** [comment]
yes, the margin is still an issue

further changes I'll make:
- Maximize views by double clicking a tab
- A tooltip when hovering over the `expand / restore` buttons
- Probably a tooltip when hovering over the view tag indicating that it can be maximized by double clicking
- different positioning of the buttons (in a maximized view I'll hide the `close` button and show the the `restore` button instead)

**Comment by thinkh** [comment]
Your suggested changes sound good to me.

**Comment by mstreit** [comment]
very nice, looks great!

File: src/layout/interfaces.ts [hunk 1]
```diff
@@ -19,7 +19,7 @@ export class LayoutContainerEvents {
   static readonly EVENT_TAB_REORDED = 'tabReorded';
   static readonly EVENT_CHANGE_ACTIVE_TAB = 'changeActiveTab';
   static readonly EVENT_MAXIMIZE = 'maximize';
-  static readonly EVENT_MINIMIZE = 'minimize';
+  static readonly RESTORE_SIZE = 'restoreSize';
```
Comment (thinkh) [inline src/layout/interfaces.ts]: Why don't you keep the `EVENT_` at the beginning of the variable name?

File: src/layout/internal/RootLayoutContainer.ts [hunk 2]
```diff
@@ -10,13 +10,45 @@ export default class RootLayoutContainer extends AParentLayoutContainer<ILayoutC
...
+
   constructor(document: Document, public readonly build: (layout: IBuildAbleOrViewLike)=> ILayoutContainer, private readonly restorer: (dump: ILayoutDump, restoreView: (referenceId: number) => IView) => ILayoutContainer) {
     super(document, {
       name: '',
       fixed: true
     });
     this.node.dataset.layout = 'root';
     this.visible = true;
+
+    this.on(LayoutContainerEvents.EVENT_MAXIMIZE, (evt) => {
```
Comment (sgratzl) [inline src/layout/internal/RootLayoutContainer.ts]: prefer: `_evt, view'

File: src/layout/internal/RootLayoutContainer.ts [hunk 3]
```diff
@@ -10,13 +10,45 @@ export default class RootLayoutContainer extends AParentLayoutContainer<ILayoutC
...
+      // maximize views
+      const view = evt.args[0];
+
+      const section = document.createElement('section');
```
Comment (sgratzl) [inline src/layout/internal/RootLayoutContainer.ts]: prefer: `this.node.ownerDocument.createElement`

File: src/layout/internal/ViewLayoutContainer.ts [hunk 4]
```diff
@@ -18,6 +18,8 @@ export default class ViewLayoutContainer extends ALayoutContainer<IViewLayoutCon
     this.node.dataset.layout = 'view';
     this.node.appendChild(view.node);
 
+    this.header.insertAdjacentHTML('beforeend', `<button type="button" title="Expand view" class="size-toggle ${this.options.fixed ? 'hidden' : ''}" aria-label="Toggle View Size"><span><i class="fa fa-expand"></i></span></button>`);
```
Comment (sgratzl) [inline src/layout/internal/ViewLayoutContainer.ts]: when the layout is fixed, I cannot maximize the view?

**Pull request merged**